dm/syncer/data_validator.go

```diff
@@ -15,23 +15,75 @@ package syncer
 
 import (
 	"context"
+	"fmt"
+	"strings"
 	"sync"
 	"time"
 
 	"github.com/go-mysql-org/go-mysql/replication"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb-tools/pkg/filter"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/shopspring/decimal"
+	"go.uber.org/atomic"
 	"go.uber.org/zap"
 
 	"github.com/pingcap/tiflow/dm/dm/config"
 	"github.com/pingcap/tiflow/dm/dm/pb"
 	"github.com/pingcap/tiflow/dm/dm/unit"
+	"github.com/pingcap/tiflow/dm/pkg/binlog"
+	"github.com/pingcap/tiflow/dm/pkg/binlog/event"
 	"github.com/pingcap/tiflow/dm/pkg/conn"
 	tcontext "github.com/pingcap/tiflow/dm/pkg/context"
 	"github.com/pingcap/tiflow/dm/pkg/log"
+	"github.com/pingcap/tiflow/dm/pkg/schema"
 	"github.com/pingcap/tiflow/dm/pkg/terror"
 	"github.com/pingcap/tiflow/dm/pkg/utils"
 	"github.com/pingcap/tiflow/dm/syncer/dbconn"
 )
 
+const (
+	checkInterval      = 5 * time.Second
+	validationInterval = 10 * time.Second
+)
+
+type validateTableInfo struct {
+	Source     *filter.Table
+	Info       *model.TableInfo
+	PrimaryKey *model.IndexInfo
+	Target     *filter.Table // target table after route
+	pkIndices  []int
```

Comment:
what is it used for?

Reply:
when generate select, used to get pk column names and values

Reply:
not sure about it, will left to todo to handle it later.

Reply:
@WizardXiao PTAL